sqlmesh/core/snapshot/evaluator.py

```diff
@@ -1606,6 +1605,66 @@ def delete(self, name: str, **kwargs: t.Any) -> None:
         self.adapter.drop_table(name)
         logger.info("Dropped table '%s'", name)
 
+    def _replace_query_for_model(
+        self,
+        model: Model,
+        name: str,
+        query_or_df: QueryOrDF,
+        render_kwargs: t.Dict[str, t.Any],
+        **kwargs: t.Any,
+    ) -> None:
+        """Replaces the table for the given model.
+
+        Args:
+            model: The target model.
+            name: The name of the target table.
+            query_or_df: The query or DataFrame to replace the target table with.
+        """
+        if (model.is_seed or model.kind.is_full) and model.annotated:
+            columns_to_types = model.columns_to_types_or_raise
+            source_columns: t.Optional[t.List[str]] = list(columns_to_types)
+        else:
+            # Source columns from the underlying table to prevent unintentional table schema changes during restatement of incremental models.
+            columns_to_types, source_columns = self._get_target_and_source_columns(
+                model, name, render_kwargs, columns_to_types=self.adapter.columns(name)
+            )
+
+        self.adapter.replace_query(
+            name,
+            query_or_df,
+            table_format=model.table_format,
+            storage_format=model.storage_format,
+            partitioned_by=model.partitioned_by,
+            partition_interval_unit=model.partition_interval_unit,
+            clustered_by=model.clustered_by,
+            table_properties=kwargs.get("physical_properties", model.physical_properties),
+            table_description=model.description,
+            column_descriptions=model.column_descriptions,
+            columns_to_types=columns_to_types,
+            source_columns=source_columns,
+        )
+
+    def _get_target_and_source_columns(
+        self,
+        model: Model,
+        table_name: str,
+        render_kwargs: t.Dict[str, t.Any],
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
+    ) -> t.Tuple[t.Dict[str, exp.DataType], t.Optional[t.List[str]]]:
+        if not columns_to_types:
+            columns_to_types = (
+                model.columns_to_types if model.annotated else self.adapter.columns(table_name)
```

Comment:
I'm a bit confused. I thought we should only use `model.columns_to_types` for `source_columns` and never for actual `columns_to_types`.

Btw, this is another reason why we should be consistent about the `columns_to_types` name throughout the code base.

Reply:
I don't think we should rely on the caller to provide the correct `columns_to_types`. I think this method should make a decision instead

Reply:
I think `model.columns_to_types` if fine but I should add a condition to this which is `if model.annotated not model.on_destructive_change.is_ignore`. Does that make sense?

Reply:
the problem with this is that this method assumes that the callers does the right thing, which is a dangerous assumption. If you **always** check `model.on_destructive_change.is_ignore` and run `adapter.columns` then you risk running `adapter.columns` more than once like in case of SCD2